RecoMET/METFilters/plugins/HFNoisyHitsFilter.cc

```diff
@@ -0,0 +1,148 @@
+  const edm::EDGetTokenT<HFRecHitCollection> hfhits_token_;
+  const double rechitPtThreshold_;
+  const std::vector<std::string> listOfNoises_;
+  const bool taggingMode_;
+  const bool debug_;
+};
+
+//
+// constructors and destructor
+//
+HFNoisyHitsFilter::HFNoisyHitsFilter(const edm::ParameterSet& iConfig)
+    : hfhits_token_(consumes<HFRecHitCollection>(iConfig.getParameter<edm::InputTag>("hfrechits"))),
+      rechitPtThreshold_(iConfig.getParameter<double>("rechitPtThreshold")),
+      listOfNoises_(iConfig.getParameter<std::vector<std::string>>("listOfNoises")),
+      taggingMode_(iConfig.getParameter<bool>("taggingMode")),
+      debug_(iConfig.getParameter<bool>("debug")) {
+  produces<bool>();
+}
+
+//
+// member functions
+//
+
+// ------------ method called on each new Event  ------------
+bool HFNoisyHitsFilter::filter(edm::StreamID, edm::Event& iEvent, const edm::EventSetup& iSetup) const {
+  using namespace edm;
+  bool pass = true;
+
+  // Calo Geometry - needed for computing E_t
+  ESHandle<CaloGeometry> pG;
+  iSetup.get<CaloGeometryRecord>().get(pG);
+  const CaloGeometry* geo = pG.product();
+
+  edm::Handle<HFRecHitCollection> theHFhits;
+  iEvent.getByToken(hfhits_token_, theHFhits);
+
+  std::vector<int> noiseBits = getNoiseBits();
+
+  //Loop over the HF rechits. If one of them has Et>X and fires one the noise bits, declare the event as bad
+  for (HFRecHitCollection::const_iterator hfhit = theHFhits->begin(); hfhit != theHFhits->end(); hfhit++) {
+    float ene = hfhit->energy();
+    float et = 0;
+    // compute transverse energy
+    const GlobalPoint& poshf = geo->getPosition(hfhit->detid());
+    float pf = poshf.perp() / poshf.mag();
+    et = ene * pf;
+    if (et < rechitPtThreshold_)
+      continue;
+    int hitFlags = hfhit->flags();
+    for (unsigned int i = 0; i < noiseBits.size(); i++) {
+      if (noiseBits[i] < 0)
+        continue;
+      if (((hitFlags >> noiseBits[i]) & 1) > 0) {
+        pass = false;
+        break;
+      }
+    }
+    if (!pass)
+      break;
+  }
+  iEvent.put(std::make_unique<bool>(pass));
+  return taggingMode_ || pass;
+}
+
+std::vector<int> HFNoisyHitsFilter::getNoiseBits() const {
+  std::vector<int> result;
+  for (unsigned int i = 0; i < listOfNoises_.size(); i++) {
+    if (listOfNoises_[i] == "HFLongShort")
+      result.push_back(HcalPhase1FlagLabels::HFLongShort);
+    else if (listOfNoises_[i] == "HFS8S1Ratio")
+      result.push_back(HcalPhase1FlagLabels::HFS8S1Ratio);
+    else if (listOfNoises_[i] == "HFPET")
+      result.push_back(HcalPhase1FlagLabels::HFPET);
+    else if (listOfNoises_[i] == "HFSignalAsymmetry")
+      result.push_back(HcalPhase1FlagLabels::HFSignalAsymmetry);
+    else if (listOfNoises_[i] == "HFAnomalousHit")
+      result.push_back(HcalPhase1FlagLabels::HFAnomalousHit);
+    else if (debug_)
+      edm::LogWarning("HFNoisyHitsFilter")
+          << "Couldn't find the bit index associated to this string: " << listOfNoises_[i];
```

Comment:
this looks like a case for an exception, because a wrong noise name string would be a bug in configuration